lib/actions-oclif/device/restart.ts

```diff
@@ -0,0 +1,183 @@
+/**
+ * @license
+ * Copyright 2016-2020 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import type { IArg } from '@oclif/parser/lib/args';
+import Command from '../../command';
+import * as cf from '../../utils/common-flags';
+import { getBalenaSdk, getCliUx, stripIndent } from '../../utils/lazy';
+import { tryAsInteger } from '../../utils/validation';
+import type {
+	BalenaSDK,
+	DeviceWithServiceDetails,
+	CurrentServiceWithCommit,
+} from 'balena-sdk';
+
+interface FlagsDef {
+	help: void;
+	service?: string[];
+}
+
+interface ArgsDef {
+	uuid: string;
+}
+
+export default class DeviceRestartCmd extends Command {
+	public static description = stripIndent`
+		Restart containers on a device.
+
+		Restart containers on a device.
+		If the --service flag is provided, then only those services' containers
+		will be restarted, otherwise all containers on the device will be restarted.
+
+		Note this does not reboot the device, to do so use instead \`balena device reboot\`.
+		`;
+	public static examples = [
+		'$ balena device restart 23c73a1',
+		'$ balena device restart 23c73a1 --service myService',
+		'$ balena device restart 23c73a1 -s myService1 -s myService2',
+	];
+
+	public static args: Array<IArg<any>> = [
+		{
+			name: 'uuid',
+			description: 'the uuid of the device to restart',
+			parse: (dev) => tryAsInteger(dev),
+			required: true,
+		},
+	];
+
+	public static usage = 'device restart <uuid>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		service: flags.string({
+			description: 'the service name(s) to restart',
+			char: 's',
+			multiple: true,
+		}),
+		help: cf.help,
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { args: params, flags: options } = this.parse<FlagsDef, ArgsDef>(
+			DeviceRestartCmd,
+		);
+
+		const balena = getBalenaSdk();
+		const ux = getCliUx();
+
+		ux.action.start('Restarting services', '', { stdout: true });
```

Comment:
* Would this be better printed to stderr? If the command was to print a JSON object with the list of device UUIDs that failed / succeeded to restart, then the JSON object could be printed to stdout, and these animations printed to stderr.
* Thinking about non-interactive usage (scripts), it should be possible to disable "ASCII animations" like this. Related issues:
  * https://github.com/balena-io/balena-cli/issues/1378
  * https://github.com/balena-io/balena-cli/issues/1172
  * https://github.com/balena-io/balena-cli/issues/1089
  * https://github.com/balena-io/balena-cli/issues/1001

  I think we should have TTY auto detection, but also a flag that overrides detection.

Reply:
> Would this be better printed to stderr?

Hmm perhaps.  I purposefully printed this to stout, based on our previous discussion about stdout/stderr, but perhaps I missunderstood at the time, or misremembered.  We should probably come up with a short policy, about output and add it to the contributing doc.

> Thinking about non-interactive usage (scripts), it should be possible to disable "ASCII animations" like this. 

According to the docs this gracefully handles ttl/non-ttl.  I don't think it supports an override flag, but I think it would be good to stick to standard implementations if we can.

https://github.com/oclif/cli-ux#cliaction

Reply:
Latest push prints spinner output to stderr.